notifications/manager.go

```diff
@@ -0,0 +1,214 @@
+func (n *Manager) Run(ctx context.Context, readyChan chan struct{},
+) error {
+
+	n.runCtx = ctx
+
+	err := n.RegisterNotifications()
+	if err != nil {
+		return err
+	}
+
+	// Notify the caller that we're ready, this means we have successfully
+	// fetched a l402 token and have subscribed to the notifications at least
+	// once.
+	close(readyChan)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// RegisterNotifications registers a new server notification stream.
+func (n *Manager) RegisterNotifications() error {
+
+	// In order to create a valid l402 we first are going to call
+	// the FetchL402 method. As a client might not have outbound capacity
+	// yet, we'll retry until we get a valid response.
+	if !n.hasL402 {
+		n.fetchL402(n.runCtx)
+	}
+
+	ctx, cancel := context.WithCancel(n.runCtx)
+
+	// We'll now subscribe to the swap server notifications.
+	notifStream, err := n.cfg.Client.SubscribeNotifications(
+		ctx, &swapserverrpc.SubscribeNotificationsRequest{},
+	)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	log.Debugf("Successfully subscribed to server notifications")
+
+	// We'll now start a goroutine that will handle all the incoming
+	// notifications.
+	go func() {
+		for {
+			notification, err := notifStream.Recv()
+			if err == nil && notification != nil {
+				log.Debugf("Received notification: %v", notification)
+				n.handleNotification(notification)
+				continue
+			}
+			log.Errorf("Error receiving "+
+				"notification: %v", err)
+
+			cancel()
+
+			// If we encounter an error, we'll
+			// try to reconnect.
+			for {
+				select {
+				case <-n.runCtx.Done():
+					return
+
+				case <-time.After(time.Second * 10):
+					log.Debugf("Reconnecting to server notifications")
+					err = n.RegisterNotifications()
+					if err != nil {
+						log.Errorf("Error reconnecting: %v", err)
+						continue
+					}
+
+					// If we were able to reconnect, we'll
+					// return.
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+// fetchL402 fetches the L402 from the server. This method will keep on
+// retrying until it gets a valid response.
+func (m *Manager) fetchL402(ctx context.Context) {
+	// Add a 0 timer so that we initially fetch the L402 immediately.
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-timer.C:
+			err := m.cfg.FetchL402(ctx)
+			if err != nil {
+				log.Warnf("Error fetching L402: %v", err)
+				timer.Reset(time.Second * 10)
+				continue
+			}
+			m.hasL402 = true
+			return
+		}
+	}
+}
+
+// handleNotification handles an incoming notification from the server,
+// forwarding it to the appropriate subscribers.
+func (n *Manager) handleNotification(notification *swapserverrpc.
+	SubscribeNotificationsResponse) {
+
+	switch notification.Notification.(type) {
+	case *swapserverrpc.SubscribeNotificationsResponse_ReservationNotification:
+		for _, fn := range n.subscribers[NotificationTypeReservation] {
+			fn(notification)
+		}
```

Comment:
I propose to add a `default` case and log an error there (unknown notification type).

This will also fix the lint warning:
```
singleCaseSwitch: should rewrite switch statement to if statement (gocritic)
```